docs: create "observability" getting started guide (#1376)
keptn/lifecycle-toolkit: docs/content/en/docs/getting-started/observability/_index.md

---
title: Standardize access to observability data
description: Learn how the Keptn Lifecycle Toolkit provides observability for Kubernetes deployments
weight: 45
---

The Keptn Lifecycle Toolkit (KLT) makes any Kubernetes deployment observable.
You can readily see why a deployment takes so long or why it fails,
even when using multiple deployment tools.
Keptn provides a Keptn application resource
that aggregates multiple Workloads into a single resource

> that can be monitored.

Reviewer: I would maybe say it introduces a concept of an application, which is natively not part of the Kubernetes and therefore is able to create a layer upon running workloads

## Install and configure KLT

The steps to install and configure KLT for this exercise are:

* [Install KLT](#install-klt)
* [Enable KLT on your cluster](#enable-klt-on-your-cluster)
* [Annotate Deployment resource](#enable-klt-on-your-cluster)

### Install KLT

To install the Keptn Lifecycle Toolkit on your cluster:

   - Be sure that your cluster includes the components discussed in
     [Prepare your cluster for KLT](../../install/k8s.md/#prepare-your-cluster-for-klt)
   - [Install and enable the Keptn Lifecycle Toolkit](../../install/install.md/#use-helm-chart)
     on your cluster using the Helm chart

If you installed KLT on your cluster for the
[Getting started with Keptn metrics](../metrics)
exercise, you do not need to re-install it for this exercise.
However, if you only installed the `metrics-operator` for that exercise,
you now need to install the full KLT.

### Enable KLT on your cluster

Follow the instructions in
[Enable KLT for your cluster](../../install/install.md/#enable-klt-for-your-cluster)
to enable KLT on your cluster.

The
[simplenode-dev-ns.yaml](https://github.com/keptn-sandbox/klt-on-k3s-with-argocd/blob/main/simplenode-dev/simplenode-dev-ns.yaml)

> file defines the `Namespace` resource for our example.

Author: I'm not sure I understand.  You don't have to enable KLT or apply annotations for metrics so yeah, the install section is a lot shorter.  I thought this was kind of parallel -- tell them the step, point to the regular docs for instructions.  And I had a link to the appropriate file in the sample which wasn't applicable for Metrics but I think is useful here.  We'll talk...

> file defines the `Namespace` resource for our example.

Reviewer 2: @StackScribe this conversation is the only outstanding thing for me before a merge.

> file defines the `Namespace` resource for our example.

Author: I refactored the beginning material a bit and am waiting to hear if the new version satisfies @thisthat .  Long term, I want the landing page of the "Install" section to have this list and we'll just point to that from any Getting Started materials.

The first GS guide for KeptnMetrics is, of course, shorter and simpler because they don't need to enable KLT or annotate the yaml file.

> file defines the `Namespace` resource for our example.

Reviewer 3: I agree to go ahead with the current version and do a follow-up to improve the overall install sections

You can create the YAML file to define the
[KeptnApp](../../yaml-crd-ref/app.md)
resource manually
but the easier approach is to let KLT create this definition for you.
This requires that you annotate all your workloads
(`Deployments`, `Stateful Sets`, `DaemonSets`, and `ReplicaSets`
as described in
[Use Keptn automatic app discovery](../../implementing/integrate.md/#use-keptn-automatic-app-discovery).

## DORA metrics

DORA metrics are an industry-standard set of measurements;
see the following for a description:

* [What are DORA Metrics and Why Do They Matter?](https://codeclimate.com/blog/dora-metrics)
* [Are you an Elite DevOps Performer?  Find out with the Four Keys Project](https://cloud.google.com/blog/products/devops-sre/using-the-four-keys-to-measure-your-devops-performance)

DORA metrics provide information such as:

* How many deployments happened in the last six hours?
* Time between deployments
* Deployment time between versions
* Average time between versions.

The Keptn Lifecycle Toolkit starts collecting these metrics
as soon as you annotate the `Deployment` resource.

TODO: Are Dora metrics collected when KLT is enabled or only

>       when the `Deployment` resource is annotated?

Author: Thanks -- done

TODO: How to observe DORA metrics without a dashboard.
      The info about observing Keptn metrics seems specific to those;
      if it's the same commands, what are the names of each DORA

>       metric to specify?

Author: Thanks -- incorported

## Using OpenTelemetry

> The Keptn Lifecycle Toolkit extends the Kubernetes
> pod scheduler to create OpenTelemetry data

Reviewer 3: To be more future proof since soon-ish we won't use anymore the scheduler.
```suggestion
The Keptn Lifecycle Toolkit extends the Kubernetes
primitives to create OpenTelemetry data
```

that connects all your deployment and observability tools
without worrying about where it is stored and where it is managed.
OpenTelemetry traces collect data as Kubernetes is deploying the changes,
which allows you to trace everything done in the context of that deployment.

* You must have an OpenTelemetry collector installed on your cluster.
  See

>   [OpenTelemetry Collector](https://www.dynatrace.com/support/help/extend-dynatrace/opentelemetry/basics/collector)

Reviewer 3: I would use the OpenTelemetry link for this
```suggestion
  [OpenTelemetry Collector](https://opentelemetry.io/docs/collector/)
```

  for more information.
* Follow the instructions in
  [OpenTelemetry observability](../../implementing/otel.md)

>   to configure Keptn accessibility to your OpenTelemetry data.

Reviewer 3: ```suggestion
  to configure where sending your OpenTelemetry data.
```

Author: done

  * Define a [KeptnConfig](../../yaml-crd-ref/config.md) resource
  that defines the URL and port of the OpenTelemetry collector.
  For our example, this is in the
  [keptnconfig.yaml](https://github.com/keptn-sandbox/klt-on-k3s-with-argocd/blob/main/setup/keptn/keptnconfig.yaml)
  file.

## Keptn metrics

You can supplement the DORA Metrics and OpenTelemetry information
with information you explicitly define using Keptn metrics.
The
[Getting started with Keptn metrics](../metrics)
exercise discusses how to define Keptn metrics.

## View the results

To start feeding observability data for your deployments
onto a dashboard of your choice,
modify either your `Deployment` or `KeptnApp` resource yaml file
to increment the version number
and commit that change to your repository.

> TODO: How to view DORA and OpenTelemetry data without a dashboard.

Author: applied

      Are the DORA metrics viewed using the same commands shown for
      Keptn metrics.

TODO: talk about the Grafana display.

The videos that go with this exercise show how the
DORA, OpenTelemetry, and Keptn metrics information

> appears on a Grafana dashboard with the

Reviewer 3: ```suggestion
appears on a Grafana dashboard with
```

[Jaeger](https://grafana.com/docs/grafana-cloud/data-configuration/metrics/prometheus-config-examples/the-jaeger-authors-jaeger/).

If you also have Jaeger extension for Grafana installed on your cluster,
you can view full end-to-end trace for everything
that happens in your deployment.

> For more information, see

Reviewer 3: I think there is a missing link here

Author: Whoops -- it's fixed

or just look at it for examples
as you implement the functionality "from scratch"
on your local Kubernetes deployment cluster.
The
[README](https://github.com/keptn-sandbox/klt-on-k3s-with-argocd/blob/main/setup/observability/README.md)
file for that repo contains useful information.

Two videos are available
to walk you through this exercise if you prefer:

* [Introducing Keptn Lifecycle Toolkit](https://youtu.be/449HAFYkUlY)
* [Use SLOs and get DORA the Native K8s way!](https://www.youtube.com/watch?v=zeEC0475SOU)

In the
[Getting started with Keptn metrics](../metrics)
exercise, you learn how to define and use Keptn metrics.
You may want to complete that exercise before doing this exercise
although that is not required.

This exercise shows how to standardize access
to the observability data for your cluster.
The steps are:

1. Install and configure the Lifecycle Toolkit on your cluster
1. Configure OpenTelemetry for the Lifecycle Toolkit
1. Create a Keptn application that aggregates your deployment workloads
1. DORA metrics
1. Using OpenTelemetry data
1. Keptn metrics

TODO: Make the list items links to the subsections.

- Increment the version number for either your Workload

>   or your application to start aggragating data.

Reviewer: ```suggestion
  or your application to start aggregating data.
```

---
title: Standardize access to observability data
description: Learn how the Keptn Lifecycle Toolkit provides observability for Kubernetes deployments
weight: 45
---

The Keptn Lifecycle Toolkit (KLT) makes any Kubernetes deployment observable.
You can readily see why a deployment takes so long or why it fails,
even when using multiple deployment tools.
Keptn introduces a concept of an application
which is an abstraction that connects multiple
Workloads belonging together.

> In other words, KLT, creates a distributed end-to-end trace

Reviewer 4: I would move this sentence above, maybe also rephrase a bit: 

The Keptn Lifecycle Toolkit (KLT) makes any Kubernetes deployment observable; in other words, it creates a distributed end-to-end trace of everything Kubernetes does in the context of a Deployment.
The advantage is that you can readily see why a deployment takes so long or why it fails,
even when using multiple deployment tools. To do so, Keptn introduces the concept of an 'application'
which is an abstraction that connects multiple
Workloads belonging together.

of everything Kubernetes does in the context of a Deployment.

The observability data is an amalgamation of the following:

- DORA metrics are collected out of the box
  when the Lifecycle Toolkit is enabled
- OpenTelemetry runs traces that show everything that happens in the Kubernetes cluster
  and can display this information with dashboard tools
  such as Grafana.
- Specific metrics that you can define to monitor
  information from all the data providers configured in your cluster.

> The Keptn Lifecycle Toolkit can provide this information

Reviewer 4: All this paragraph untile line 34 repeats what is written in 7-14

as you implement the functionality "from scratch"
on your local Kubernetes deployment cluster.
The
[README](https://github.com/keptn-sandbox/klt-on-k3s-with-argocd/blob/main/setup/observability/README.md)
file for that repo contains useful information.

Two videos are available
to walk you through this exercise if you prefer:

- [Introducing Keptn Lifecycle Toolkit](https://youtu.be/449HAFYkUlY)
- [Use SLOs and get DORA the Native K8s way!](https://www.youtube.com/watch?v=zeEC0475SOU)

In the
[Getting started with Keptn metrics](../metrics)
exercise, you learn how to define and use Keptn metrics.
You may want to complete that exercise before doing this exercise
although that is not required.

This exercise shows how to standardize access
to the observability data for your cluster.
The steps are:

1. [Install and enable]( #install-and-enable-klt)
   the Lifecycle Toolkit on your cluster
1. [Integrate the Lifecycle Toolkit with your applications](#integrate-the-lifecycle-toolkit-with-your-applications)
1. [DORA metrics](#dora-metrics)
1. [Using OpenTelemetry](#using-opentelemetry)
1. [Keptn metrics](#keptn-metrics)
1. [View the results](#view-the-results)

## Install and enable KLT

> To install and enable the Keptn Lifecycle Toolkit on your cluster:

Reviewer 4: One user was asking us whether the KLT can be installed in a separate cluster, unsure if this is the right spot to stress it out or if we should do it in the installation. But I would have the info that KLT communicates with the k8s scheduler to stop deployment from proceeding until tasks/evaluations pass, thus it needs to be installed in the cluster you want to monitor.

Author: I saw that question and I agree -- it doesn't belong in the Getting Started Guide but in the main Installation section.  I will do a separate PR for this.

1. Follow the instructions in
   [Install the Keptn Lifecycle Toolkit](../../install/install.md/#use-helm-chart)
   to install KLT on your cluster using the Helm chart

   If you installed KLT on your cluster for the
   [Getting started with Keptn metrics](../metrics)
   exercise, you do not need to re-install it for this exercise.
   However, if you only installed the `metrics-operator` for that exercise,
   you now need to install the full KLT.

1. Follow the instructions in
   [Enable KLT for your cluster](../../install/install.md/#enable-klt-for-your-cluster)
   to enable KLT on your cluster
   by annotating the `Namespace` resource..
   See the
   [simplenode-dev-ns.yaml](https://github.com/keptn-sandbox/klt-on-k3s-with-argocd/blob/main/simplenode-dev/simplenode-dev-ns.yaml)
   file for an example

1. Run the following command to ensure that your Kuberetes cluster
   is ready to complete this exercise:

   ```shell
   kubectl get pods -n keptn-lifecycle-toolkit-system
   ```

   You should see pods for the following components:
   - certificate-operator (or another cert manager)
   - lifecycle-operator
   - scheduler

>    - otel-collector

Reviewer 3: We don't install this and users might have it in a different namespace. I would remove from the list

Author: done

Author: done
